### PR TITLE
Refactor some more string handling in lib

### PR DESCRIPTION
**Pull request by Geod24**
Just a collection of commit aiming to use more arrays in Lib.
Also found a few odd casts, and some non-DRY `removeExt`, which also prompted me to remove some overloads in FileName now that most places use the D array API.

### CODE REVIEW

File: src/dmd/root/filename.d [hunk 1]
```diff
@@ -95,12 +95,6 @@ nothrow:
      * Returns:
      *  true if absolute path name.
      */
-    extern (C++) static bool absolute(const(char)* name) pure @nogc
-    {
-        return absolute(name.toDString);
-    }
-
-    /// Ditto
```
Comment (thewilsonator) [inline src/dmd/root/filename.d]: Headers?

**Review by thewilsonator — APPROVED** [review]
Otherwise looks good.

**Comment by thewilsonator** [comment]
```
ERROR: dmd\link.d(353): Error: none of the overloads of `ext` are callable using argument types `(const(char)*)`, candidates are:

dmd\root\filename.d(149):        `dmd.root.filename.FileName.ext(const(char)[] str)`

dmd\root\filename.d(185):        `dmd.root.filename.FileName.ext()`

dmd\link.d(921): Error: function `dmd.root.filename.FileName.absolute(const(char)[] name)` is not callable using argument types `(const(char)*)`

dmd\link.d(921):        cannot pass argument `cmd` of type `const(char)*` to parameter `const(char)[] name`

```

**Comment by thewilsonator** [comment]
```
ERROR: dmd/libelf.d(77): Error: function void dmd.libelf.LibElf.addObject(const(char)* module_name, const(ubyte[]) buffer) does not override any function, did you mean to override void dmd.lib.Library.addObject(const(char)[] module_name, const(ubyte[]) buf)?
```

File: src/dmd/libmscoff.d [hunk 2]
```diff
@@ -312,13 +312,13 @@ final class LibMSCoff : Library
         om.base = cast(ubyte*)buf;
         om.length = cast(uint)buflen;
         om.offset = 0;
-        const(char)* n = global.params.preservePaths ? module_name : FileName.name(module_name); // remove path, but not extension
-        om.name = n.toDString();
+        const n = global.params.preservePaths ? module_name : FileName.name(module_name); // remove path, but not extension
```
Comment (thewilsonator) [inline src/dmd/libmscoff.d]: `n` is otherwise unused, just assign it directly to `om.name`

Comment (Geod24) [reply src/dmd/libmscoff.d]: Right

File: src/dmd/libelf.d [hunk 3]
```diff
@@ -74,26 +74,26 @@ final class LibElf : Library
...
-    override void addObject(const(char)* module_name, const ubyte[] buffer)
+    override void addObject(const(char)[] module_name, const ubyte[] buffer)
     {
-        if (!module_name)
-            module_name = "";
         static if (LOG)
         {
-            printf("LibElf::addObject(%s)\n", module_name);
+            printf("LibElf::addObject(%s)\n",
+                   cast(int)module_name.length, module_name.ptr);
```
Comment (MoonlightSentinel) [inline src/dmd/libelf.d]:
```suggestion
                   cast(int) module_name.length, module_name.ptr);
```

Ditto below

Comment (Geod24) [reply src/dmd/libelf.d]: Hum, sorry, by the time I checked the D style this was merged. Actually forgot this was a rule. Will keep it in mind for future PRs.

Comment (MoonlightSentinel) [reply src/dmd/libelf.d]: No big deal, just improves the readability IMHO

**Pull request merged**